src/components/controllers/slot-controller.ts

```diff
@@ -0,0 +1,189 @@
+ * as well as much custom code from our original SlottedController.
+ *
+ * This updated controller handles:
+ * - Default
+ *   - Ability to test if a named slot is present/populated
+ *   - Ability to test if the default slot is present/populated
+ *   - Get a list of all available slots
+ * - Optional
+ *   - ShadowShift
+ *     - Moving all named slots to ShadowDOM
+ *     - Moving all default slots to ShadowDOM
+ *
+ * @todo: Make the ShadowShift functionality able to target specific slots only.
+ * @see https://github.com/shoelace-style/shoelace/blob/next/src/internal/slot.ts
+ * @example
+ * // Do NOT move LightDOM elements.
+ * slots = new SlotController(
+ *   this, // The host element.
+ *   false // To shift or not to shift LightDom nodes to ShadowDOM.
+ * );
+ * @example
+ * // Move LightDOM elements to the ShadowDOM.
+ * slots = new SlotController(
+ *   this, // The host element.
+ *   true // To shift or not to shift LightDom nodes to ShadowDOM.
+ * );
+ */
+export class SlotController implements ReactiveController {
+  host: ReactiveControllerHost & Element;
+  hostEl: ReactiveElement;
+  shadowShift: boolean;
+  slotNames: string[] = [];
+  defaultSlot: boolean;
+  namedSlots: boolean;
+
+  constructor(host: ReactiveControllerHost & Element, shadowShift?: boolean) {
+    (this.host = host).addController(this);
+    this.hostEl = this.host as unknown as ReactiveElement;
+    this.shadowShift =
+      shadowShift && typeof shadowShift === 'boolean' ? shadowShift : false;
+    this.handleSlotChange = this.handleSlotChange.bind(this);
+  }
+
+  private hasDefaultSlot() {
+    return [...this.host.childNodes].some(node => {
+      if (node.nodeType === node.TEXT_NODE && node.textContent!.trim() !== '') {
+        return (this.defaultSlot = true);
+      }
+
+      if (node.nodeType === node.ELEMENT_NODE) {
+        const el = node as HTMLElement;
+        if (!el.hasAttribute('slot')) {
+          return (this.defaultSlot = true);
+        }
+      }
+
+      return (this.defaultSlot = false);
+    });
+  }
+
+  private hasNamedSlot(name: string) {
+    return (this.namedSlots =
+      this.host.querySelector(`:scope > [slot="${name}"]`) !== null);
+  }
+
+  test(slotName?: string | undefined) {
+    const slot = slotName ? slotName : '';
+    return slot === '' ? this.hasDefaultSlot() : this.hasNamedSlot(slot);
+  }
+
+  hostConnected() {
+    this.host.shadowRoot!.addEventListener('slotchange', this.handleSlotChange);
+    this.slotNames = this.getSlots();
+    setTimeout(() => this.moveSlots(), 0);
+  }
+
+  hostDisconnected() {
+    this.host.shadowRoot!.removeEventListener(
+      'slotchange',
+      this.handleSlotChange
+    );
+  }
+
+  handleSlotChange(event: Event) {
+    const slot = event.target as HTMLSlotElement;
+
+    if (
+      (this.defaultSlot && !slot.name) ||
+      (slot.name && this.slotNames.includes(slot.name))
+    ) {
+      this.host.requestUpdate();
+    }
+  }
```

Comment:
I can verify that adding new content to the slot doesn't bring that new element into the ShadowDOM. 

![image](https://user-images.githubusercontent.com/21905/154157262-cd060a16-c983-4549-9a39-1d5aff76d76c.png)

This `handleSlotChange` method will need some additional tweaking and testing.

![image](https://user-images.githubusercontent.com/21905/154157420-6b2af2ba-3df2-4e88-920b-33d413b58a09.png)

Reply:
This is now fixed. Just needed to add `moveSlots()` to the slot change handler.